Workflow output definition

bentsherman opened the pull request:
This PR is a prototype for defining an output schema for a Nextflow pipeline. See https://github.com/nextflow-io/nextflow/issues/4669 and https://github.com/nextflow-io/nextflow/issues/4670 for original discussions.

The meta workflow that we are targeting is:

fetchngs -> rnaseq -> differentialabundance

In other words, we want to eliminate the manual curation of samplesheets between each pipeline. To do this, the output schema should "mirror" the params schema, it should describe the outputs as a collection of samplesheets.

Here is the tree of pipeline outputs for the fetchngs `test` profile:
```
/custom/user-settings.mkfg
/fastq/DRX024467_DRR026872.fastq.gz
/fastq/DRX026011_DRR028935_1.fastq.gz
/fastq/DRX026011_DRR028935_2.fastq.gz
/fastq/ERX1234253_ERR1160846.fastq.gz
/fastq/SRX10940790_SRR14593545_1.fastq.gz
/fastq/SRX10940790_SRR14593545_2.fastq.gz
/fastq/SRX11047067_SRR14709033.fastq.gz
/fastq/SRX17709227_SRR21711856.fastq.gz
/fastq/SRX17709228_SRR21711855.fastq.gz
/fastq/SRX6725035_SRR9984183.fastq.gz
/fastq/SRX9315476_SRR12848126_1.fastq.gz
/fastq/SRX9315476_SRR12848126_2.fastq.gz
/fastq/SRX9504942_SRR13055517_1.fastq.gz
/fastq/SRX9504942_SRR13055517_2.fastq.gz
/fastq/SRX9504942_SRR13055518_1.fastq.gz
/fastq/SRX9504942_SRR13055518_2.fastq.gz
/fastq/SRX9504942_SRR13055519_1.fastq.gz
/fastq/SRX9504942_SRR13055519_2.fastq.gz
/fastq/SRX9504942_SRR13055520_1.fastq.gz
/fastq/SRX9504942_SRR13055520_2.fastq.gz
/fastq/SRX9626017_SRR13191702_1.fastq.gz
/fastq/SRX9626017_SRR13191702_2.fastq.gz
/fastq/md5/DRX024467_DRR026872.fastq.gz.md5
/fastq/md5/DRX026011_DRR028935_1.fastq.gz.md5
/fastq/md5/DRX026011_DRR028935_2.fastq.gz.md5
/fastq/md5/ERX1234253_ERR1160846.fastq.gz.md5
/fastq/md5/SRX17709227_SRR21711856.fastq.gz.md5
/fastq/md5/SRX17709228_SRR21711855.fastq.gz.md5
/fastq/md5/SRX6725035_SRR9984183.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055517_1.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055517_2.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055518_1.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055518_2.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055519_1.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055519_2.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055520_1.fastq.gz.md5
/fastq/md5/SRX9504942_SRR13055520_2.fastq.gz.md5
/fastq/md5/SRX9626017_SRR13191702_1.fastq.gz.md5
/fastq/md5/SRX9626017_SRR13191702_2.fastq.gz.md5
/metadata/DRR026872.runinfo_ftp.tsv
/metadata/DRR028935.runinfo_ftp.tsv
/metadata/ERR1160846.runinfo_ftp.tsv
/metadata/GSE214215.runinfo_ftp.tsv
/metadata/GSM4907283.runinfo_ftp.tsv
/metadata/SRR12848126.runinfo_ftp.tsv
/metadata/SRR13191702.runinfo_ftp.tsv
/metadata/SRR14593545.runinfo_ftp.tsv
/metadata/SRR14709033.runinfo_ftp.tsv
/metadata/SRR9984183.runinfo_ftp.tsv
/samplesheet/id_mappings.csv
/samplesheet/multiqc_config.yml
/samplesheet/samplesheet.csv
```

From what I can tell, the `samplesheet.csv` contains all of the metadata, including the file paths, MD5 checksums, and id mappings. So the samplesheet and the fastq files comprise the essential outputs and everything else is duplication.

The initial output schema basically describes this samplesheet in a similar manner to the `input_schema.json` file. This particular output schema should closely resemble the `input_schema.json` for nf-core/rnaseq.

What I'd to do from here is collect feedback on this approach -- what else is needed to complete the output schema for this pipeline? Then we can think about how to operationalize it in Nextflow -- should Nextflow automatically generate the samplesheet from the schema? how does the schema interact with the publish mechanism? how to collect metadata which normally can't be published directly but only through files?

bentsherman committed:
Add workflow output schema

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

ewels commented on output_schema.json:
```diff
@@ -0,0 +1,125 @@
+                        "type": "string"
+                    }
+                },
+                "required": ["sample", "fastq_1"]
```
I take it this is a way to set anything that's _not_ `required` as being an optional output?

bentsherman replied:
This is the convention used by the `schema_input.json` for rnaseq. Personally I would prefer to specify `"required": true` for each required param, but for now I'm just following the current convention

ewels replied:
It's more than just nf-core convention, it's [JSON schema syntax](https://json-schema.org/understanding-json-schema/reference/object#required).

I'm not 100% clear on whether we need to mirror metadata like this in the _output_ schema purely because it's present in the input schema of another pipeline. It doesn't have any real semantic value here if it's not describing the outputs. The other pipeline will still validate fine as long as these outputs are here, they don't need to be set as required here.

I'd kind of prefer to use this field to have semantic value for the outputs if we could. For example, if we know that this pipeline will _always_ produce certain files then we could mark them as `required`. Then we know that any downstream pipeline can rely on that. Anything not tagged as required would be an optional output and could be dealt with accordingly.

bentsherman replied:
fair enough, I don't feel strongly about it either way

bentsherman replied:
> if we know that this pipeline will always produce certain files then we could mark them as required. Then we know that any downstream pipeline can rely on that. Anything not tagged as required would be an optional output and could be dealt with accordingly.

This ☝️ is exactly what I intended by putting the `required` fields in the output schema. Just wanted to clarify because I think I didn't fully understand what you were saying at first.

So I guess all of these fields should be marked "required" because It looks like they are all always produced. It would be nicer to instead only mark the optional ones as "optional", but I guess we have to conform to the JSON schema standard.

evanfloden commented:
Clarifying the schema added in this PR is the output equivalent of the samplesheet schema `schema_input.json` [here](https://github.com/nf-core/rnaseq/blob/master/assets/schema_input.json)? You have it as `input_schema.json` above.

I had considered this schema to be defining one of the inputs/outputs of a pipeline. Whereas the `nextflow_schema.json` in the [base](https://github.com/nf-core/rnaseq/blob/master/nextflow_schema.json) directory of the repo defines all the possible inputs. Is that correct?

bentsherman commented:
@evanfloden this output schema is like the `nextflow_schema.json` with the `schema_input.json` embedded. So it lists all of the outputs but each samplesheet output has it's own schema embedded instead of in a separate file for simplicitly.

@adamrtalbot At the very least, this output schema should be used to validate any samplesheets that are produced, and allow external tools like Seqera Platform to inspect a workflow's expected outputs e.g. for the purpose of chaining pipelines.

What isn't clear to me yet is whether the output schema can be used to automate the generation of the samplesheet.

bentsherman commented:
I think that could work. As long as the channel emits maps (or records once we support record types properly), generating the samplesheet is trivial.

pditommaso commented:
This looks going in the right direction. One thing I found awful in the current schema is *JSON schema* that's totally unreadable. Wonder if we should not look into a different system more human friendly

pditommaso commented:
Possible alternatives 
* [YAML schema](https://asdf-standard.readthedocs.io/en/1.0.1/schemas/yaml_schema.html)
* [Yamale](https://github.com/23andMe/Yamale)

drpatelh commented:
My biggest concern with this is how unwieldy that file is going to get when we go to defining an output schema for a simple pipeline like fetchngs to rnaseq. This is why I was suggesting we try and incorporate the publishing logic and output file definition at the module/subworkflow/workflow level and then combine them somehow rather than having one single massive file.

I also suspect there will still need to be some sort of "conversion" layer or plugin that can take this output schema file to generate custom csvs/jsons etc which can be used as input downstream for other pipelines. Ideally, this plugin can be invoked outside of the pipeline context.

bentsherman committed:
Convert output schema to YAML

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add output definition

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

ewels commented:
> but each samplesheet output has it's own schema embedded instead of in a separate file for simplicitly.

I don't think that we should do this, it breaks how JSON schema validation works. The beauty of using the standard is that very many platforms and libraries use the syntax in the same way. You have a parsed object in memory (be it `params` of the contents of a sample sheet, doesn't really matter) and you validate it against a schema.

If we start merging sample sheet schema inside output schema, we can no longer use this for validation. We would have to validate the output files with subsets of the schema, and validate the list of output files with a subset of the schema. If you have to break the schema down to use it, it becomes custom and a lot less useful imho. Separate files is undoubtably more verbose, but it's also much more portable.

This is why the `nextflow_schema.json` for params refers to the path to a separate schema file for any given files, rather than embedding that logic within.

ewels commented:
@pditommaso YAML is fine (and Ben's YAML conversion here hopefully is a lot easier to read), but my strong preference is to stick with as-close-to-as-possible JSON Schema _syntax_.

To clarify, that JSON Schema can be written in YAML (or toml, or really any format), as long as it's laid out with the structure and keywords of JSON schema. The benefit of using it is that there are about a [bazillion different implementations](https://json-schema.org/implementations) so it just works everywhere.

In contrast, the [Yamale](https://github.com/23andMe/Yamale) syntax you linked to seems to by a Python tool with it's own schema syntax, so every part of our toolchain would need to build its own parser and validation library for that syntax.

The [YAML Schema](https://asdf-standard.readthedocs.io/en/1.6.0/schemas/yaml_schema.html) you linked to seems to still be valid JSON Schema, just in YAML format and with a couple of extra keys. That would still work with any JSON Schema implementation, so that'd be fine. But I'm not sure that we're doing anything complex enough to need those extra keywords to be honest.

bentsherman committed:
Simplify output DSL, samplesheet generation

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
I converted the JSON schema to YAML just to see what it looks like, and it is indeed much simpler. If the JSON schema can be used with YAML "schemas" just the same, that seems like the best approach to me, even for the `nextflow_schema.json`.

bentsherman commented:
I also added a prototype for the workflow output DSL (see https://github.com/nextflow-io/nextflow/pull/4784). It allows you to define an arbitrarily nested directory structure with `path()`, then publish process outputs with `select()` using a process selector and the standard publish options. It is, in my opinion, stupid simple

Another idea I considered was being able to select channels from the top-level workflow emits, but that is slightly more complicated to implement (and adds some boilerplate to the pipeline code) whereas I found I could get the job done with just the process outputs.

I thought about having some DSL method like `index <source-channel> <filename>` which could collect metadata records from a channel and write them to a file. It's actually pretty trivial to do with Groovy, fetchngs was just doing it in a roundabout way, so I simplified some things in the pipeline code.

bentsherman commented:
At this point, the output DSL is concerned only with mapping process outputs to a directory structure. Where output schemas could come in is as an optional refinement to describe the structure of specific files:
```groovy
select 'SRA_TO_SAMPLESHEET', pattern: 'samplesheet.csv', schema: 'schema_samplesheet.json'
select 'SRA_TO_SAMPLESHEET', pattern: 'id_mapping.csv', schema: 'schema_mapping.json'
```

So it's still up to the user to generate the output file, and they might even be able to use the same output schema to do it (like Adam's `toSamplesheet()` example). But the above definition can be used by external tools and users to understand the structure of workflow outputs without running the pipeline.

Given this example, I agree with @ewels that it makes more sense to keep the schema for each file separate. I'm imaging a nextflow command to generate some kind of global schema from this output definition (i.e. by the pipeline developer before a version release) for use by external tools.

bentsherman committed:
Flatten output definition

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix bug

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add default publish mode

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
See https://github.com/nf-core/rnaseq/pull/1227 for a similar prototype with rnaseq. It is not for the faint of heart

bentsherman committed:
Add `toSamplesheet()` method to generate samplesheet

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Move output file schemas to separate files

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add `schema` option to validate published outputs against a schema

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented on modules/local/sra_to_samplesheet/main.nf:
```diff
@@ -1,75 +1,83 @@
 
-process SRA_TO_SAMPLESHEET {
-    tag "$meta.id"
+/**
+ * Save a list of records to a samplesheet file.
+ *
+ * @param records
+ * @param path
+ */
+def toSamplesheet(List<Map> records, Path path) {
+    def lines = []
+    lines << records.first().keySet().collect{ '"' + it + '"'}.join(",")
+    records.each { record ->
+        lines << record.values().collect{ '"' + it + '"'}.join(",")
+    }
+
+    path.text = lines.join('\n')
+}
```
@adamrtalbot here is your `toSamplesheet()` method 😄 

This can be extended to select/encode fields based on a schema, and support other output formats. I envision it being added to nf-validation (and eventually Nextflow as a standard library function) to complement `fromSamplesheet()`.

bentsherman commented on main.nf:
```diff
@@ -83,6 +83,28 @@ workflow {
+        }
+
+        path('samplesheet') {
+            select 'SRA_TO_SAMPLESHEET', pattern: 'samplesheet.csv', schema: 'assets/schema_samplesheet.yml'
+            select 'SRA_TO_SAMPLESHEET', pattern: 'id_mappings.csv', schema: 'assets/schema_mappings.yml'
```
This `schema` option could be used to validate output files matching this selector. The user should generate these files in their pipeline code, but this is trivial to do with something like `toSamplesheet()`.

bentsherman commented on output.yml:
I feel like this top-level schema file could be automatically generated from the output DSL (e.g. with a nextflow command)

mahesh-panchal commented:
I guess the schema would be optional right, a nice support? The learning curve for Nextflow is already high to new-comers and asking for JSON understanding on top should be nice to have rather than a must have. 

How would publishing using task variables and other workflow variables look like?

Does a workflow user have the ability to change these settings some how, or are they fixed? Can they be augmented, for example if I want to publish some files also to another folder?

subwaystation commented:
What if I don't have a samplesheet as input? As in https://github.com/nf-core/pangenome.

mahesh-panchal commented:
Did I misunderstand and this is just for validating output, and not replacing `publishDir`?

bentsherman commented:
> I guess the schema would be optional right, a nice support?

Yes the output schema is optional. The goal is primarily to replace `publishDir` and secondarily provide a way to validate outputs

> How would publishing using task variables and other workflow variables look like?

The output DSL will be scoped to the script or workflow block (not sure which one yet), so it will be able to use any variables in that scope. Task variables aren't supported since the publishing is decoupled from the individual tasks

> Does a workflow user have the ability to change these settings some how, or are they fixed? Can they be augmented, for example if I want to publish some files also to another folder?

Because the output definition is part of the pipeline code, it can't really be changed beyond the use of params. You can provide a param to toggle some output, but if the user wants to add something new, they'll need to fork the pipeline

> What if I don't have a samplesheet as input?

The output schema can be used with many different file formats (JSON, YAML, CSV), not just samplesheets, and if your pipeline input can't be easily validated with JSON schema then you don't have to use it, it's optional

subwaystation commented:
I have a simple FASTA.gz as input. Should be easily doable, if I understand correctly.

bentsherman commented:
If it's just a single input file then you can probably provide any metadata through params, so there would be no need for an index file (i.e. samplesheet) or schema to validate it

mahesh-panchal commented:
> > How would publishing using task variables and other workflow variables look like?
> 
> The output DSL will be scoped to the script or workflow block (not sure which one yet), so it will be able to use any variables in that scope. Task variables aren't supported since the publishing is decoupled from the individual tasks
> 

Doesn't this take away a major feature like publishing files to a folder based on sample name? I think there are quite a few examples where some field of meta for example is used in the publish path.

bentsherman commented:
> Doesn't this take away a major feature like publishing files to a folder based on sample name?

I think I have seen this pattern before, though I couldn't find an example of it in rnaseq.

It is a consequence of decoupling the publishing from the task execution. We might be able to recover it in #302 by allowing the path to reference channel items, e.g. given a channel of files with metadata, publish the file to a path based on the meta id, but not sure what that syntax would look like.

bentsherman commented:
@adamrtalbot good point, with channel selectors we could do something like this:
```groovy
path( "results" ) {
    select NFCORE_RNASEQ.out.bam, saveAs: { meta, bam -> "${meta.id}/bam/${bam.name}" }
}
```

The only thing is, I was imagining the selected channel would just provide paths, but if they provide tuples/records with files and metadata, it's not obvious how the file elements are being pulled out of the tuple.

mahesh-panchal commented:
>  it's not obvious how the file elements are being pulled out of the tuple.

Isn't this how it is now? Only `path` types are published. `val`, `env`, etc are ignored.

bentsherman committed:
Replace process selectors with channel selectors

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
Since there is a lot of support for the channel selectors, but most of the discussion is here, I closed the other PR and migrated this one to the channel selectors.

I found that topics weren't really needed for fetchngs. We'll see what happens with rnaseq

bentsherman commented:
> Isn't this how it is now? Only path types are published. val, env, etc are ignored.

@mahesh-panchal yes, because of how process outputs are declared, Nextflow knows which tuple elements are paths and collects them accordingly

Since a generic channel doesn't have such a declaration, for now I think we can follow @adamrtalbot 's suggestion and traverse whatever data structure the channel spits out:
- if it's a path or path list, publish it
- if it's a tuple, traverse each element
  - if the element is a path or path list, publish it
- otherwise raise an error

Once we have support for record types and better type inference in the language, we'll be able to infer the incoming type and, if it's a record type, scan the type definition for path elements. Much more robust but not a blocker for the quick n dirty solution

bentsherman committed:
Use channel topics to reduce channel propagation

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Temporary workaround for topic operator bug

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented on workflows/sra/main.nf:
```diff
@@ -54,6 +54,7 @@ workflow SRA {
     SRA_RUNINFO_TO_FTP
         .out
         .tsv
+        // .topic('runinfo-tsv')
```
This currently doesn't work because it causes the pipeline to hang. Only happens when the `topic` operator is inserted into the middle of an operator chain. Still investigating in the Nextflow PR

bentsherman committed:
Apply updates from upstream

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

maxulysse commented on main.nf:
```diff
@@ -83,6 +91,32 @@ workflow {
+    }
+
+    'samplesheet' {
+        from NFCORE_FETCHNGS.out.samplesheet // , schema: 'assets/schema_samplesheet.yml'
+        from NFCORE_FETCHNGS.out.mappings // , schema: 'assets/schema_mappings.yml'
```
```suggestion
        from NFCORE_FETCHNGS.out.samplesheet // , schema: 'assets/schema_samplesheet.yml'
        from NFCORE_FETCHNGS.out.mappings    // , schema: 'assets/schema_mappings.yml'
```

maxulysse replied:
We need to respect Harshil's alignment

bentsherman replied:
that's not an actual comment it's just not implemented yet 😆

maxulysse commented on main.nf:
```diff
@@ -83,6 +91,32 @@ workflow {
+output {
+    directory params.outdir, mode: params.publish_dir_mode
+
+    'fastq' {
+        from 'fastq'
+    }
```
ok, so the first `fastq` is the path, and the second is the topic.
Can it be a bit more explicit?

Thinking something like this at least:
```suggestion
    'fastq/' {
        from 'fastq'
    }
```

bentsherman replied:
you can specify a trailing slash if you want

maxulysse replied:
I'd like that, it makes the path more explicit, but I'd rather even have a `path` and a `topic` specified somewhere, not to be too confused, I think it's better to be a bit more explicit

bentsherman replied:
That's fair. This case is simple enough to be confusing, but if you used a regular emit instead of a topic, or you had a more complex directory structure like in rnaseq, it would be clearer.

I did propose calling it `fromTopic` to help denote it as a topic, but Paolo was in favor of not having too many different keywords. We could revisit this

bentsherman commented:
@adamrtalbot totally fair. fetchngs is simple enough that you could do it without topics, and you are certainly free to scrap the topics in the final implementation. I used them here as an exercise to show how to access those intermediate channels without adding them to the emits. In practice, users will be able to use any mixture of emits / topics based on their preference.

bentsherman committed:
Update using latest syntax

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented on workflows/sra/main.nf:
```diff
@@ -181,14 +182,25 @@ workflow SRA {
     sra_metadata    = ch_sra_metadata
     versions        = ch_versions.unique()
+
+    publish:
+    ch_fastq                    >> 'fastq/'
+    ASPERA_CLI.out.md5          >> 'fastq/md5/'
+    SRA_FASTQ_FTP.out.md5       >> 'fastq/md5/'
+    SRA_RUNINFO_TO_FTP.out.tsv  >> 'metadata/'
+    ch_versions_yml             >> 'pipeline_info/'
+    ch_samplesheet              >> 'samplesheet/'
+    ch_mappings                 >> 'samplesheet/'
+    ch_sample_mappings_yml      >> 'samplesheet/'
```
By default, the "topic" name will be used as the publish path, which makes fetchngs really simple. No need to define any rules in the output DSL, just the base directory and publish mode, then all of these channels will be published exactly as it says.

These names don't have to be paths, they can also be arbitrary names which you would then use in the output DSL to customize publish options for that name. I'll demonstrate this with rnaseq. You can think of the names as "topics" if you want, but at this point I'm not even using topics under the hood because they aren't necessary.

bentsherman committed:
Rename output -> publish

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
minor updates

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
Folding into #312
